File: Exercism/JAVA/blackjack/src/main/java/Blackjack.java
import java.util.HashMap;

public class Blackjack {

  public int parseCard(String card) {
    HashMap<String, Integer> cards = new HashMap<String, Integer>();

    cards.put("ace", 11);
    cards.put("eight", 8);
    cards.put("two", 2);
    cards.put("nine", 9);
    cards.put("three", 3);
    cards.put("ten", 10);
    cards.put("four", 4);
    cards.put("jack", 10);
    cards.put("five", 5);
    cards.put("queen", 10);
    cards.put("six", 6);
    cards.put("king", 10);
    cards.put("seven", 7);
    cards.put("other", 0);

    return cards.get(card);
  }

  public boolean isBlackjack(String card1, String card2) {
    throw new UnsupportedOperationException(
        "Please implement the Blackjack.isBlackjack method");
  }

  public String largeHand(boolean isBlackjack, int dealerScore) {
    throw new UnsupportedOperationException(
        "Please implement the Blackjack.largeHand method");
  }

  public String smallHand(int handScore, int dealerScore) {
    throw new UnsupportedOperationException(
        "Please implement the Blackjack.smallHand method");
  }

  // FirstTurn returns the semi-optimal decision for the first turn, given the
  // cards of the player and the dealer. This function is already implemented
  // and does not need to be edited. It pulls the other functions together in a
  // complete decision tree for the first turn.
  public String firstTurn(String card1, String card2, String dealerCard) {
    int handScore = parseCard(card1) + parseCard(card2);
    int dealerScore = parseCard(dealerCard);

    if (20 < handScore) {
      return largeHand(isBlackjack(card1, card2), dealerScore);
    } else {
      return smallHand(handScore, dealerScore);
    }
  }
}

File: Exercism/ruby/blackjack/blackjack.rb
module Blackjack
  def self.parse_card(card)
    case card
    when "ace" then 11
    when "two" then 2
    when "three" then 3
    when "four" then 4
    when "five" then 5
    when "six" then 6
    when "seven" then 7
    when "eight" then 8
    when "nine" then 9
    when "ten", "jack", "queen", "king" then 10
    else 0
    end
  end

  def self.card_range(card1, card2)
    sum = parse_card(card1) + parse_card(card2)
    case sum
    when 4..11 then "low"
    when 17..20 then "high"
    when 12..16 then "mid"
    when 21 then "blackjack"
    end
  end

  def self.first_turn(card1, card2, dealercard)
    sum = parse_card(card1) + parse_card(card2)
    range = card_range(card1, card2)
    if %w[high mid].include?(range) || (sum == 21 && dealercard == "ace")
      "S"
    elsif sum <= 11
      "H"
    elsif card1 == "ace" && card2 == "ace"
      "P"
    elsif sum == 21 && (dealercard != "ace" || parse_card(dealercard) != 10)
      "W"
    end
  end
end
